src/consensus/aft/test/committable_suffix.cpp

```diff
@@ -0,0 +1,860 @@
+      }
+    }
+  }
+
+  aft::ChannelStubProxy::MessageList best_only;
+  for (const auto& [node_id, pair] : best)
+  {
+    const auto& [idx, it] = pair;
+    best_only.push_back(*it);
+  }
+
+  messages = std::move(best_only);
+}
+
+#define TEST_DECLARE_NODE(N) \
+  ccf::NodeId node_id##N(#N); \
+  auto store##N = std::make_shared<AllSigsStore>(node_id##N); \
+  TRaft r##N( \
+    ConsensusType::CFT, \
+    std::make_unique<AllSigsAdaptor>(store##N), \
+    std::make_unique<aft::LedgerStubProxy>(node_id##N), \
+    std::make_shared<aft::ChannelStubProxy>(), \
+    std::make_shared<aft::StubSnapshotter>(), \
+    nullptr, \
+    nullptr, \
+    cert, \
+    std::make_shared<aft::State>(node_id##N), \
+    nullptr, \
+    nullptr, \
+    nullptr, \
+    request_timeout, \
+    election_timeout, \
+    election_timeout); \
+  initial_config[node_id##N] = {}; \
+  nodes[node_id##N] = &r##N; \
+  auto channels##N = channel_stub_proxy(r##N);
+
+/**
+  Summary of the behaviour this test aims to explore (for implementation reasons
+  the indices in the actual test don't match those in this comment, but the
+  logical relationships are the same):
+
+  - Produce an initial state where A is primary in term 1 of a 5-node network,
+    has mixed success replicating its entries. Local commit index is marked by
+    []. True commit index is the highest of these, [1.2].
+    A:  1.1 [1.2] 1.3
+    B: [1.1] 1.2  1.3
+    C: [1.1] 1.2
+    D: [1.1]
+    E: [1.1]
+
+  - Intuitively, B and C are responsible for persisting 1.2, although they don't
+    know this locally.
+
+  - Node A dies. Node B becomes primary, and creates some tail junk that no-one
+    else sees. It makes no commit progress. Crucially this is committable (as
+    all of these indices are), so B is reluctant to discard the entries in term
+    2, and 1.3, though all _could_ be discarded.
+    A:  1.1 [1.2] 1.3 (DEAD)
+    B: [1.1] 1.2  1.3  2.4  2.5
+    C: [1.1] 1.2
+    D: [1.1]
+    E: [1.1]
+
+  - Node C doesn't hear from B for long enough, and becomes primary. Its
+    AppendEntries are regularly lost. Eventually a heartbeat ("I'm at 1.2")
+    reaches Node B, which responds with a NACK ("I don't accept that tail").
+    Node C responds to that NACK with a better AppendEntries, which _should_
+    bring Node B back to the committed level, and forward to wherever C is.
+    However, under the current implementation, this AppendEntries may
+    actually cause B to roll back further than is safe, losing the committed
+    state.
+    C->B: AE [1.2, 1.2)
+    B->C: AER NACK 1.1 (B's commit index)
+    C->B: AE [1.1, 1.1) (Large entries mean this is only a partial catchup)
+    B: Rolls back to 1.1
+
+    B: [1.1]
+    C: [1.1] 1.2
+    D: [1.1]
+    E: [1.1]
+
+  - At this point a committed index (1.2) is no longer present on a majority of
+    nodes. While C is unlikely to advertise it (fancy election rules mean
+    they're waiting for commit at 1.2) and will continue to share it, its
```

Comment:
Not sure I understand that bit. If it was practically committed, it must have been committable (ie. signed) and so if C participates there's a good chance it will win. This is all moot because the election could resolve without C of course, as rightly pointed out in the next line.

Reply:
I'll reword this. I was trying to say that the user-visible commit may not show the loss of persistence (ie - if C survives, then it _will_ share `1.2`, and that will be committed again, and the commit that a user could see won't move backwards). I think there may be space for an interesting distinction between the actual commit, and the user-visible commit, since the fancy election state lets a node know while have an inaccurate/lagging commit index? But as you say, that's all moot since there's still a fundamental persistence hole - _if_ C loses an election here, then we've lost persistence.

Reply:
Ah I see, advertise it to the user, I'd read that as advertise it in the election process. Yes, there is a distinction not just between "persisted by f+1 nodes" and "commit on leader", but also "commit as advertised by node to user".